### ISSUE TITLE
switch Formatters to use Template Toolkit

### ISSUE THREAD

**Issue by morgajel** [CityGenerator/CityGenerator#97]
I should be able to replace a good chunk of fugly code into templates.

**Comment by three18ti** [comment]
I'd like to work on this, looks like it might be similar to issue #65 so maybe I could work on the two together.

I was wondering if you can provide some further information / clarification as to what you are looking for?

For instance, looking at EconomyFormatter.pm, I see you're initializing a "$content" string then appending to the string, if i understand correctly, you'd like to replace the string mangling with a template?  You could then feed printBusinesses or printResources a hash (or it would generate the hash and call the "template evaluate method") with the required parameters for your template, then return the formatted string.

Is this along the lines of what you're thinking or am I way off?

**Comment by morgajel** [comment]
yup- you're pretty close to what I'm looking for (oh, and I closed #65 since it was a duplicate).

You can look at the multiflaggenerator's print_worldnav() subroutine for a good example of how I'm implementing the Template Toolkit (presuming you're not already familiar with it). EconomyFormatter is printing a list in both of its subroutines, so you can probably use the [foreach directive](http://www.template-toolkit.org/docs/manual/Directives.html#section_FOREACH) to loop through those in a sane manner.

If you focus on just a single method for EconomyFormatter (for starters), I can give you feedback. Once we're sure we're on the same page you can go nuts on all of them :)

**Comment by three18ti** [comment]
Ok, I did a quick mockup of the templates and implemented the processing of the template for printResources.  I think I need to understand better what is happening in printBusinesses, but I think it's just a matter of munging the data so we're only presenting the template with the data it is going to present.  

Also, as `if (%hash)` is the same as `if (scalar keys %hash > 0 )` I rewrote the if statement in printBusinesses (mostly because I was trying to understand what was going on)

https://github.com/three18ti/CityGenerator/blob/develop/lib/EconomyFormatter.pm

That said, I don't know if the following is worth raising another issue for the discussion, but how do you install and test?  There's quite a few dependencies it seems, I didn't know if you had a better way to install the needed modules than just grepping through the directory for "use" statements?  I was working on:

```
grep -ri '^use ' ./* | grep -v 'strict\|warnings\|vars\|base\|lib' | perl -F'\s+' -anE '$modules{$F[1]} = 1; END { map { s/;//; say } keys %modules}' | cpanm
```

but had to run before I could commit it.  

I'd like to be able to test code changes, but was having some trouble getting everything set up.  Open to any suggestions.

Finally...  The general wisdom of /r/perl and #perl (irc.perl.org) is that XML::Simple is a collection of bugs flying in loose formation.  As XML::Simple has been the source of a number of bugs in my programs (and endless frustration debugging said problems), I can only agree.  I think XML::Toolkit is a much better choice, and objections to discussing replacing XML::Simple with XML::Toolkit (probably does deserve a separate issue).

ooohh... and on the topic of XML... how hard would it be to be able to understand json instead of xml docs?

open to thoughts on my resource_template, as this was just a quick mockup I think eventually all the templates would be in their own template files, but that's not necessarily a requirement.

the businesses_template is still in progress, I need to think on that a bit more, but I'm open to thoughts.

**Comment by three18ti** [comment]
Ok, actually, just tried it on another vm.  that grep kinda worked... still had to install Lingua::EN::Conjugate and XML::LibXML::SAX (which required libxml2-dev, I think I only had to manually install because I didn't have the prereq. library installed) .

./runatest.sh failed a few tests, 

```
    #   at tests//TestSummaryFormatter.pm line 25.
    #          got: 'Kanhall is a small town in the Britkan Province with a .'
    #     expected: 'Kanhall is a small town in the Britkan Province with a normal population of around 1,138.'
```

But I don't think that is my fault (is it?... should have tried running the test before I made any changes... but had trouble finding/installing dependencies...).

How can I run the test for just the economy?  (is there one?...  Looking at the tests directory now, unless I'm just missing it, I don't think there is one... I'm a bit bleary eyed after a 12 hour shift so excuse me if I've overlooked something obvious)

**Comment by morgajel** [comment]
Correct, not your fault- I fixed that earlier today.

**Comment by morgajel** [comment]
ok, it's taking me a moment to parse everything you've written; comments will come shortly.

**Comment by morgajel** [comment]
> Ok, I did a quick mockup of the templates and implemented the processing of the template for printResources. I  think I need to understand better what is happening in printBusinesses, but I think it's just a matter of munging the data so we're only presenting the template with the data it is going to present.

So the good news is I've recently Axed printBusinesses altogether for the time being (per ticket #93) so for now you only need to worry about printResources.

> Also, as if (%hash) is the same as if (scalar keys %hash > 0 ) I rewrote the if statement in printBusinesses (mostly because I was trying to understand what was going on)

They very well may be the same, I don't know off the top of my head... as long as the tests pass it's all good.

> https://github.com/three18ti/CityGenerator/blob/develop/lib/EconomyFormatter.pm

The only thing that sticks out is that I'd like to see the templates loaded from the Templates/ directory rather than inline- that way it can be used by other implementations in the future. Otherwise they look good:

> That said, I don't know if the following is worth raising another issue for the discussion, but how do you install and test? There's quite a few dependencies it seems, I didn't know if you had a better way to install the needed modules than just grepping through the directory for "use" statements? I was working on:
> 
> grep -ri '^use ' ./\* | grep -v 'strict|warnings|vars|base|lib' | perl -F'\s+' -anE '$modules{$F[1]} = 1; END { map { s/;//; say } keys %modules}' | cpan
> but had to run before I could commit it.

As it stands right now, I don't have a good solution; I need to come up with a better method.

> I'd like to be able to test code changes, but was having some trouble getting everything set up. Open to any suggestions.

I'm happy to work with you over IM to get it up and running; I suggest that we document it in the wiki for future use.

> Finally... The general wisdom of /r/perl and #perl (irc.perl.org) is that XML::Simple is a collection of bugs flying in loose formation. As XML::Simple has been the source of a number of bugs in my programs (and endless frustration debugging said problems), I can only agree. I think XML::Toolkit is a much better choice, and objections to discussing replacing XML::Simple with XML::Toolkit (probably does deserve a separate issue).

I'm not opposed to getting rid of XML::Simple in general, but that will be a massive project since it's at the very core of every single Generator. I'd be willing to consider it for the next major point release if we can come up with some strong metrics proving the next solution is more efficient (which shouldn't be dfifficult).

> ooohh... and on the topic of XML... how hard would it be to be able to understand json instead of xml docs?

I actually considered this in the past and created the benchmark branch for that explicit purpose- JSON.pm was demonstrability slower than the other 5 backend libraries (speaking of which, just added #132). If this weren't the case I'd consider it, although I do not look forward to the idea of re-migrating all that data from xml to json and still having it be human readable.

> open to thoughts on my resource_template, as this was just a quick mockup I think eventually all the templates would be in their own template files, but that's not necessarily a requirement.

agreed- as mentioned above, toss it in Templates and I'll be content.

> the businesses_template is still in progress, I need to think on that a bit more, but I'm open to thoughts.

As mentioned earlier- scrap it, I just killed that section because it was redundant. I may bring it back later with different content, but for now it's not needed.

> Ok, actually, just tried it on another vm. that grep kinda worked... still had to install Lingua::EN::Conjugate and XML::LibXML::SAX (which required libxml2-dev, I think I only had to manually install because I didn't have the prereq. library installed) .
> 
> ./runatest.sh failed a few tests,
> 
> ```
> #   at tests//TestSummaryFormatter.pm line 25.
> #          got: 'Kanhall is a small town in the Britkan Province with a .'
> #     expected: 'Kanhall is a small town in the Britkan Province with a normal population of around 1,138.'
> ```
> 
> But I don't think that is my fault (is it?... should have tried running the test before I made any changes... but had trouble finding/installing dependencies...).

Yes, this was a bug caused by me impatiently pulling a fix for #90 that wasn't ready. I've since pushed a fix for it back to develop.

> How can I run the test for just the economy? (is there one?... Looking at the tests directory now, unless I'm just missing it, I don't think there is one... I'm a bit bleary eyed after a 12 hour shift so excuse me if I've overlooked something obvious)

You are dead on; the current test system is an all or nothing setup- usually I'll just comment out lines in test/runtests.pl and run ./runtest.sh full. It's non-optimal and should be improved.

**Comment by three18ti** [comment]
> So the good news is I've recently Axed printBusinesses altogether for the time being (per ticket #93) so for now you only need to worry about printResources.

Ok, cool.  I don't think printResources is covered by your test suite...  I introduced an erroneous character to  lib/EconomyFormatter.pm

```
root@raring64:~/CityGenerator# perl -Tc lib/EconomyFormatter.pm
syntax error at lib/EconomyFormatter.pm line 4, near "a#!/usr/bin/perl -wT
```

But all the tests pass... (I pulled in all your changes to my fork).

Same thing with lib/NPCFormatter.pm

Looks like EnvironmentFormatter is covered though (well, I get a syntax error, but no tests fail).  I'm looking for the low hanging fruit, maybe you could point me at a few formatters that are easy and covered by the test suite?  I might just take a crack at EnvironmentFormatter anyway.

> As it stands right now, I don't have a good solution; I need to come up with a better method.
> 
> You are dead on; the current test system is an all or nothing setup- usually I'll just comment out lines in test/runtests.pl and run ./runtest.sh full. It's non-optimal and should be improved.

You should check out Dist::Zilla, it provides a number of features, AutoPrereqs, etc. as well as a framework for running tests.  It might take a little work to implement but I think it would go a long way to making getting up and running easier.

**Comment by morgajel** [comment]
Interesting.. not sure what's up with that error- the modules aren't intended to be run individually, and it looks like it's choking on the package stanza.

I just finished up the GovtGenerator's printMilitary subroutine- that's a nice fleshy target for templating.

I'll add Dist::Zilla investigation to the issues list.

I've been  all over the board the last few days; before you invest too much time in any more code, make sure you pull the latest version of develop. I'd hate for you to get hung up on merges if you can avoid it.

**Comment by three18ti** [comment]
> They very well may be the same, I don't know off the top of my head... as long as the tests pass it's all good. 

I was having a related conversation with a co worker and worked up this proof-of-concept    https://gist.github.com/three18ti/6199085

> Interesting.. not sure what's up with that error- the modules aren't intended to be run individually, and it looks like it's choking on the package stanza.

Well, I introduced the error by making a syntax error.  My point was I'm not convinced my code is actually covered by / passes all the tests (since a syntax error in lib/EconomyFormatter.pm doesn't cause the test to fail).

> I just finished up the GovtGenerator's printMilitary subroutine- that's a nice fleshy target for templating.

Ok, cool I'll have a run at that.

> I've been all over the board the last few days; before you invest too much time in any more code, make sure you pull the latest version of develop

Good point.  I try to remember to fetch / merge from upstream before starting for the day.

**Comment by three18ti** [comment]
> https://github.com/three18ti/CityGenerator/blob/develop/lib/GovtFormatter.pm#L147

This seems to pass the tests, I'd really like to see a "before and after" though to verify the results are as expected...

Again, inline for testing, but will break it out into Templates/ once I'm convinced it does what I think it does.

Also, since all of the methods essentially require calling $tt->prove($template, $city, $output), do you have a centralized "utilities" module where I could stick a "parse template" method?  it just seems like typing Template->new all over the place is unnecessary.

**Comment by morgajel** [comment]
GenericGenerator is intended to eventually be an abstract class that the others will use (once I make the leap to OO). set_seed and roll_from_array are both examples similar to what you're talking about.

**Comment by three18ti** [comment]
Just wanted to circle back around to this, I'd like to dig into this, but my concern is without a way to verify my work , I spend time making the changes then come to find out it's got some fundamental flaw.  I'm just not certain that the tests would cover my changes (earlier I intentionally broke one of the formatters and did not get errors when I ran the test)

Thoughts?

Thanks

**Comment by morgajel** [comment]
If you check in your changes to your fork, I can take a look and give you feedback.

**Comment by three18ti** [comment]
Hey,

So this commit is where I rewrote the printResources function to use a template:
https://github.com/three18ti/CityGenerator/commit/b7b73fbf1ec1e520346977ddbe00ba581dd7a0f0

And this commit is where I rewrote the printMilitary function to use a template:
https://github.com/three18ti/CityGenerator/commit/7cfc8c8ce5677775e9750e7f0eb54e82212079e3

At the moment I'm still using inline templates as I find them easier to debug (only requiring editing one file instead of two or three.  As we have previously discussed, long term I'll be putting these into a separate template file (probably one per function).

Thanks for the feedback!

**Comment by morgajel** [comment]
ok, finally got a chance to look over it- looking good. @braddoro is curently working on #69 which may stomp your changes for resources, and I recently made changes to printMilitary, so you may want to re-merge my develop into your branch and account for any new changes- otherwise it's looking very good!

**Comment by morgajel** [comment]
moving this out of the 2.2 release to give @three18ti more time to work on it.

**Comment by morgajel** [comment]
no longer needed as we're moving to python in megacosm.
